analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/RestrictDeserializedTypesBase.cs

```diff
@@ -26,4 +29,166 @@ public abstract class RestrictDeserializedTypesBase : SymbolicRuleCheck
     protected const string MessageFormat = "{0}";
     private const string RestrictTypesMessage = "Restrict types of objects allowed to be deserialized.";
     private const string VerifyMacMessage = "Serialized data signature (MAC) should be verified.";
+
+    private static readonly KnownType[] FormattersWithBinder = new[]
+    {
+        KnownType.System_Runtime_Serialization_Formatters_Binary_BinaryFormatter,
+        KnownType.System_Runtime_Serialization_NetDataContractSerializer,
+        KnownType.System_Runtime_Serialization_Formatters_Soap_SoapFormatter
+    };
+    private static readonly KnownType JavaScriptSerializer = KnownType.System_Web_Script_Serialization_JavaScriptSerializer;
+    private static readonly KnownType LosFormatter = KnownType.System_Web_UI_LosFormatter;
+    private static readonly KnownType[] TypesWithDeserializeMethod = FormattersWithBinder.Append(JavaScriptSerializer).ToArray();
+
+    private readonly Dictionary<ISymbol, SyntaxNode> additionalLocationsForSymbols = new();
+    private readonly Dictionary<IOperation, SyntaxNode> additionalLocationsForOperations = new();
+
+    protected abstract bool IsBindToTypeMethod(SyntaxNode methodDeclaration);
+    protected abstract bool IsResolveTypeMethod(SyntaxNode methodDeclaration);
+    protected abstract bool ThrowsOrReturnsNull(SyntaxNode methodDeclaration);
+    protected abstract SyntaxToken GetIdentifier(SyntaxNode methodDeclaration);
+
+    protected override ProgramState PostProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        var operation = context.Operation.Instance;
+        if (operation.Kind == OperationKindEx.ObjectCreation)
+        {
+            return operation.Type.IsAny(FormattersWithBinder)
+                ? state.SetOperationConstraint(operation, SerializationConstraint.Unsafe)
+                : ProcessOtherSerializerCreations(state, operation.ToObjectCreation());
+        }
+        else if (operation.AsAssignment() is { } assignment)
+        {
+            if (ProcessBinderAssignment(state, assignment) is { } binderProcessedState)
+            {
+                return binderProcessedState;
+            }
+            else if (AdditionalLocation(state, assignment.Value) is { } methodDeclaration
```

Comment:
Please add some explanation why this is needed. It's not trivial.
Mention an example as well (e.g. `var formatter = new BinaryFormatter { Binder = binder };`)